### PR TITLE
Fix protocol to solve #121 and #125

### PR DESCRIPTION
**Pull request by njeisecke**
Please see the commit messages for details.

### CODE REVIEW

File: singleapplication_p.cpp [hunk 1]
```diff
@@ -376,29 +405,35 @@ void SingleApplicationPrivate::readInitMessageHeader( QLocalSocket *sock )
...
     info.msgLen = msgLen;
 
-    if ( sock->bytesAvailable() >= (qint64) msgLen ){
-        readInitMessageBody( sock );
-    }
+    writeAck( sock );
 }
 
-void SingleApplicationPrivate::readInitMessageBody( QLocalSocket *sock )
+qint64 SingleApplicationPrivate::isDataComplete( QLocalSocket *sock )
```
[the remark this answers is not in the document]
Comment (njeisecke) [reply singleapplication_p.cpp]: Yes, of course ;-)

**Review by Shatur — APPROVED** [review]
LGTM code-wise. @nioncode, @dail8859, could you test it?

File: singleapplication_p.cpp [hunk 2]
```diff
@@ -263,20 +263,46 @@ bool SingleApplicationPrivate::connectToPrimary( int msecs, ConnectionType conne
 #endif
     writeStream << checksum;
 
-    // The header indicates the message length that follows
+    return writeConfirmedMessage( static_cast<int>(msecs - time.elapsed()), initMsg );
+}
+
+void SingleApplicationPrivate::writeAck( QLocalSocket *sock ) {
```
Comment (Shatur) [inline singleapplication_p.cpp:269]: Why are we passing a socket to a function, can we use `socket` class member?

Comment (njeisecke) [reply singleapplication_p.cpp:269]: The `socket` member is only for the client use case. writeAck is invoked by the server in response to client requests. IMHO mixing both server and client functionality in the same class is not optimal.

**Comment by nioncode** [comment]
Will test this soon and get back to you, hopefully today.

**Comment by itay-grudev** [comment]
Looks good to me. I tested it on Linux.

**Review by itay-grudev — APPROVED** [review]

**Comment by nioncode** [comment]
This looks great, I just tested it on Windows 10 and Linux (NixOS) and every started instance was correctly recognized. Thanks @njeisecke !

One thing that I want to point out though: messages are now sent in a blocking way (i.e. they are sent and then the code blocks until the `ack` is received). Is this a problem @itay-grudev ?

**Comment by itay-grudev** [comment]
@nioncode I believe that is alright, as the `waitForReadyRead` is constrained within the specified timeout. I believe that design is acceptable.

**Comment by itay-grudev** [comment]
@njeisecke Thank you for you contribution. What is your name so I can add you to my CHANGELOG?

**Comment by njeisecke** [comment]
Thank you all for maintaining this library, reviewing and accepting this change. "Nils Jeisecke" that would be.

**Pull request merged**